### PR TITLE
API gateway form encoding support

### LINKED ISSUE #294: Accept and form encoding from the API gateway
The API gateway should accept and transform form encoded data to JSON and pass it on to the function/event.  This makes building some web front-ends easier

- [ ] Bug
- [ ] Feature
- [x] Enhancement

## Complexity

- [ ] Low - Simple enhancement or bug fix, no architectural changes or refactoring
- [x] Medium - Change requires some thought, but is relatively isolated
- [ ] High - Significant architectural change or large refactor

## Impact

- [ ] Low - Annoyance, but does not impact business or functionality
- [x] Medium - Issue can be worked around, but is causing pain
- [ ] High - Blocker

## Your Environment

* Dispatch CLI version (or git commit):
* Dispatch Chart version (or image[s] tag):
* Operating System and version:
* Kubernetes version (and distribution):
* Etc (any other useful environmental information):

### PR DESCRIPTION
**Pull request by rjew**
Fixes #294 
* API gateway now accepts content-types: json & form encoding
* Form encoded data is transformed to json before passing it to the function/event
* Updated tests to use content type header and added new tests to test content type functionality

### CODE REVIEW

File: charts/kong/dispatch-transformer/handler.lua [hunk 1]
```diff
@@ -84,15 +89,34 @@ local function substitute_payload(conf, result)
...
-    result[conf.substitute.input] = json_data
+  local header = ngx.req.get_headers()
+  if is_json_body(header) then
+    local ok, json_data = parse_json(data)
+    if ok then
+      ngx.log(ngx.DEBUG, "request body is json")
+      result[conf.substitute.input] = json_data
+    else
+      ngx.log(ngx.DEBUG, "request body is not json")
+      result[conf.substitute.input] = data
```
Comment (berndtj) [inline charts/kong/dispatch-transformer/handler.lua]: This seems wrong.  Shouldn't we return an HTTP error if the content type is application/json, but the payload doesn't match that?

Comment (berndtj) [reply charts/kong/dispatch-transformer/handler.lua]: I guess that's just continuing existing behavior, but it seems wrong

File: charts/kong/dispatch-transformer/handler.lua [hunk 2]
```diff
@@ -84,15 +89,34 @@ local function substitute_payload(conf, result)
...
+    end
+  elseif is_form_urlencoded_body(header) then
+    local form_urlencoded_data, err = ngx.req.get_post_args()
+    if not err then
+      ngx.log(ngx.DEBUG, "request body is x-www-form-urlencoded")
+      result[conf.substitute.input] = form_urlencoded_data
+    else
+      ngx.log(ngx.DEBUG, "request body is not x-www-form-urlencoded: " .. err)
```
Comment (berndtj) [inline charts/kong/dispatch-transformer/handler.lua]: Again, this seems wrong.

File: charts/kong/dispatch-transformer/handler.lua (hunk 1 above)
Comment (rjew) [reply charts/kong/dispatch-transformer/handler.lua]: Yeah that's a good point. I was just extending the existing behavior. I'll look into throwing an HTTP error when the payload doesn't match the content type or when the content type is not supported or specified.

File: charts/kong/dispatch-transformer/handler.lua (hunk 2 above)
Comment (kars7e) [reply charts/kong/dispatch-transformer/handler.lua]: looks like error is returned below that? Or is this comment outdated?

Comment (rjew) [reply charts/kong/dispatch-transformer/handler.lua]: Yeah that comment is outdated. I returned the error right below that line.

**Comment by rjew** [comment]
So it seems to be failing the e2e test when creating a function with default secret. I tested it locally and it seems to be working fine. It's probably timing out too early before the function can finish creating since the later test that executes the given function passes.

**Pull request merged**